text/0000-client-side-buildpack-registry.md

```diff
@@ -0,0 +1,302 @@
+
+Here's an example directory structure:
+
+```
+1
+├── a
+└── b
+2
+├── aa
+└── ab
+3
+├── a
+│   ├── abc
+│   └── acd
+└── b
+    ├── bcd
+    └── bed
+fo
+├── ob
+│   ├── fooball
+│   └── foobar
+└── oc
+    └── foocal
+```
+
+The following ids are reserved by Windows, so they aren't allowed as valid ids:
+
+* nul
+* con
+* prn
+* aux
+* com1
+* com2
+* com3
+* com4
+* com5
+* com6
+* com7
+* com8
+* com9
+* lpt1
+* lpt2
+* lpt3
+* lpt4
+* lpt5
+* lpt6
+* lpt7
+* lpt8
+* lpt9
+
+
+The file will contain minified JSON for each buildpack. Multiple entries will exist in a file split by a newline. This strikes a balance between human redable, easy to parse, and minimizing the diffs for new updates.
+
+An entry will have the following structure:
+
+```
+{
+  "buildpacks" : [
+    {
+      "ns" : "<string>",
+      "name": "<string>",
+      "version" : "<string",
+      "digest" : "<string>",
+      "yanked" : <boolean>,
+      "addr" : "<string>",
+    }
+  ],
+}
+```
+
+The `buildpacks` fields are defined as follows:
+
+* `ns` - can represent a set or organization of buildpacks.
+* `name` - an identifier that must be unique within a namespace.
+* `version` - the version of the buildpack (must match the version in the `buildpack.toml` of the buildpack)
+* `digest` - the image ID of the OCI image that represents the buildpack (used for validation)
+* `yanked` - whether or not the buildpack has been removed from the registry
+* `addr` - the address of the image stored in a Docker Registry (ex. `"docker.io/jkutner/lua"`)
+
+An example of what this may look like for a single buildpack file:
+```
+{"ns":"heroku","name":"ruby","version":"0.1.0","digest":"a9d9038c0cdbb9f3b024aaf4b8ae4f894ea8288ad0c3bf057d1157c74601b906","yanked":false,"addr":"docker.io/hone/ruby-buildpack:0.1.0"}
+{"ns":"heroku","name":"ruby","version":"0.2.0","digest":"2560f05307e8de9d830f144d09556e19dd1eb7d928aee900ed02208ae9727e7a","yanked":false,"addr":"docker.io/hone/ruby-buildpack:0.2.0"}
+{"ns":"heroku","name":"ruby","version":"0.2.1","digest":"74eb48882e835d8767f62940d453eb96ed2737de3a16573881dcea7dea769df7","yanked":false,"addr":"docker.io/hone/ruby-buildpack:0.2.1"}
+{"ns":"heroku","name":"ruby","version":"0.3.0","digest":"8c27fe111c11b722081701dfed3bd55e039b9ce92865473cf4cdfa918071c566","yanked":false,"addr":"docker.io/hone/ruby-buildpack:0.3.0"}
+```
+
+*Note:* id is the combination of two fields, `ns` and `name`. The `/` will be replaced by a `_` in the filename. For example:
+
+```
+he
+└── ro
+    └── heroku_java
+```
+
+### Manipulating the Index
+
+In general, the file will be append only for add. This will add new entries through `pack publish-buildpack`.
+
+When performing a `pack yank-buildpack`, it will rewrite the corresponding line by adjusting the `yanked` field to `true`.
+
+When performing a `pack yank-buildpack --undo`, it will rewrite the corresponding line by adjusting the `yanked` field to `false`.
+
+### Squashing the Index
```

Comment:
Why is this broadly necessary? It feels like shallow clones would be sufficient here?

Reply:
I don't have a strong opinion here. This was based on advise from the crates.io folks. We can dig into why they found this necessary.

Reply:
Trust in Rust.  No need to research.

Reply:
Indeed. I found this from an [internal rust post](https://internals.rust-lang.org/t/cargos-crate-index-upcoming-squash-into-one-commit/8440).

> Performing an incremental update of a shallow checkout is very expensive as a server operation. The CocoaPods project on GitHub [has historically run into problems](https://thenextweb.com/dd/2016/03/08/cocoapods-popular-basically-took-five-github-servers/) with this strategy. Effectively this isn’t a scalable solution on large scale.